
```{r eval=FALSE}


# conda activate voyager
library(Seurat)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(ggrepel)
library(ggpubr)
library(ggrastr)
library(patchwork)
theme_set(theme_cowplot())
set.seed(12345)

# co-expression network analysis packages:
library(WGCNA)
library(hdWGCNA)
library(igraph)
library(harmony)

# set random seed for reproducibility
set.seed(12345)

# optionally enable multithreading
enableWGCNAThreads(nThreads = 8)


setwd('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/revision/hdWGCNA')
fig_dir <- 'figures/'
data_dir <- 'data/'

# load seurat object
seurat_obj <- readRDS(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/revision/data/harmony_annotated_integration.rds')


# re-load hdWGCNA seurat object 
seurat_obj <- readRDS(file = paste0(data_dir, 'harmony_annotated_hdWGCNA.rds'))

```

write the modules to a table:

```{r eval=FALSE}

modules <- GetModules(seurat_obj) %>% subset(module != 'grey')
write.csv(modules, file=paste0(data_dir, 'modules.csv'), quote=FALSE, row.names=FALSE)

```

run hdWGCNA 

```{r eval=FALSE}

# subset by neuronal cells only:
seurat_obj <- subset(seurat_obj, cell_type %in% c('MHb-Neuron', 'LHb-Neuron', 'PHb-Neuron') & Group %in% c('Nurr2c', 'GFP'))

# get the umap from cell_meta:
umap <- seurat_obj@meta.data[,c('neuron_UMAP_1', 'neuron_UMAP_2')]
rownames(umap) <- rownames(seurat_obj@meta.data)

# set UMAP
seurat_obj@reductions$neuron_umap <- Seurat::CreateDimReducObject(
  embeddings = as.matrix(umap),
  key="neuronUMAP",
  assay="RNA"
)


source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')

color_df <- data.frame(
  colour = as.character(cluster_colors),
  group = names(cluster_colors)
)

seurat_obj$annotation <- droplevels(seurat_obj$annotation)
p<- PlotEmbedding(
    seurat_obj,
    group.by = 'annotation',
    raster_dpi = 400,
    raster_scale=0.5, point_size=0.5,
    label=TRUE,
    plot_theme =  NoLegend() + umap_theme(),
    reduction = 'neuron_umap',
    color_df = color_df
)


pdf(paste0(fig_dir, 'neuronumap.pdf'), height=3, width=4)
p
dev.off()

# subset by neuronal cells only:
seurat_obj <- subset(seurat_obj, cell_type %in% c('MHb-Neuron') & Group %in% c('Nurr2c', 'GFP'))
seurat_obj$annotation <- droplevels(seurat_obj$annotation)
seurat_obj$Sample <- droplevels(seurat_obj$Sample)

seurat_obj <- SetupForWGCNA(
  seurat_obj,
  gene_select = "fraction",
  fraction = 0.05,
  #group.by = 'cell_type',
  wgcna_name = "MHb_metacell"
)
length(GetWGCNAGenes(seurat_obj))

# seurat_obj <- MetacellsByGroups(
#   seurat_obj = seurat_obj,
#   group.by = c('cell_type', "Sample", "Group"),
#   k = 50, # k=25 # this was used on the first try
#   max_shared=40,
#   ident.group = 'cell_type',
#   reduction='harmony',
#   target_metacells=250
# )

seurat_obj <- MetacellsByGroups(
  seurat_obj = seurat_obj,
  group.by = c('annotation', "cell_type","Sample", "Group"),
  k = 25, # k=25 # this was used on the first try
  max_shared=15,
  ident.group = 'annotation',
  reduction='harmony',
  target_metacells=100,
  min_cells = 25
)

seurat_obj <- NormalizeMetacells(seurat_obj)

mobj <- GetMetacellObject(seurat_obj)
table(mobj$cell_type, mobj$Group)
#table(mobj$annotation, mobj$Group)

seurat_obj <- SetDatExpr(
  seurat_obj,
  group_name = "MHb-Neuron",
  group.by = "cell_type"
)

# test the soft power threshold
seurat_obj <- TestSoftPowers(seurat_obj)

plot_list <- PlotSoftPowers(seurat_obj)

#assemble with patchwork
pdf(paste0(fig_dir, 'softpower2.pdf'), width=12, height=8)
wrap_plots(plot_list, ncol=2)
dev.off()


seurat_obj <- ConstructNetwork(
    seurat_obj, 
    #soft_power=5,
    tom_name='MHb_metacell', 
    overwrite_tom=TRUE
)

# plot the dendrogram
pdf(paste0(fig_dir, "dendro5.pdf"),height=3, width=6)
PlotDendrogram(seurat_obj, main='MHb hdWGCNA Dendrogram')
dev.off()


modules <- GetModules(seurat_obj)
table(modules$module)

subset(modules, gene_name == 'Nr4a2')
subset(modules, gene_name == 'Gabra2')
subset(modules, gene_name == 'Sox5')



################################################################################
# Module overlap analysis between old modules and new
################################################################################

seurat_old <- readRDS('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/harmony_annotated_scWGCNA.rds')


library(GeneOverlap)

group1 <- 'new'
group2 <- 'old'

m1 <- GetModules(seurat_old, wgcna_name='MHb')
m2 <- GetModules(seurat_obj)

mods1 <- levels(m1$module); mods1 <- mods1[mods1 != 'grey']
mods2 <- levels(m2$module); mods2 <- mods2[mods2 != 'grey']

genome.size <- nrow(seurat_obj)


# run overlaps between module gene lists and DEG lists:
overlap_df <- do.call(rbind, lapply(mods1, function(cur_mod1){
  cur_m1_genes <- m1 %>% subset(module == cur_mod1) %>% .$gene_name
  cur_overlap_df <- do.call(rbind, lapply(mods2, function(cur_mod2){
    print(paste0(cur_mod1, ' ', cur_mod2))
    cur_m2_genes <- m2 %>% subset(module == cur_mod2) %>% .$gene_name
    cur_overlap <- testGeneOverlap(newGeneOverlap(
        cur_m1_genes,
        cur_m2_genes,
        genome.size=genome.size
    ))
    c(cur_overlap@odds.ratio, cur_overlap@pval, cur_overlap@Jaccard, length(cur_overlap@intersection))
  })) %>% as.data.frame
  colnames(cur_overlap_df) <- c('odds_ratio', 'pval', 'Jaccard', 'size_intersection')
  cur_overlap_df$m1 <- cur_mod1
  cur_overlap_df$m2 <- mods2

  # module color:
  #cur_overlap_df$color <- modules %>% subset(module == cur_mod) %>% .$color %>% unique
  cur_overlap_df
}))

overlap_df$m1 <- factor(as.character(overlap_df$m1), levels=mods1)
overlap_df$m2 <- factor(as.character(overlap_df$m2), levels=mods2)

# re-order
tmp <- unlist(lapply(mods2, function(cur_mod2){
  cur <- subset(overlap_df, m2==cur_mod2)
  which(cur$odds_ratio == max(cur$odds_ratio))
}))

overlap_df$m2 <- factor(as.character(overlap_df$m2), levels=rev(mods2[order(tmp)]))

# adjust for multiple comparisons:
overlap_df$fdr <- p.adjust(overlap_df$pval, method='fdr')

# significance level:
overlap_df$Significance <- gtools::stars.pval(overlap_df$fdr)
overlap_df$Significance <- ifelse(
  overlap_df$Significance == '.', '',
  overlap_df$Significance
)

# plot the results as a heatmap:
maxval <- 50
plot_df <- overlap_df
plot_df$odds_ratio <- ifelse(plot_df$odds_ratio > maxval, maxval, plot_df$odds_ratio)
plot_df$textcolor <- ifelse(plot_df$odds_ratio > 0.7*maxval, 'white', 'black')


p <- plot_df %>%
  ggplot(aes(x=m1, y=m2, fill=odds_ratio)) +
  geom_tile() +
  geom_text(label=plot_df$Significance, vjust = 0.72, color=plot_df$textcolor) +
  scale_fill_gradient(low='white', high='blue') +
  RotatedAxis() +
  labs(fill = 'Odds ratio') +
  theme(
    panel.border = element_rect(fill=NA, color='black', size=1),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    #axis.title.x = element_blank(),
    #axis.title.y = element_blank(),
    #axis.text.x = element_blank(),
    #axis.ticks.x = element_blank(),
  #  axis.text.y = element_blank(),
    # axis.ticks.y = element_blank(),
    plot.margin=margin(0,0,0,0)
  ) +
  #coord_equal() +
  ylab('New') +
  xlab('Old')


pdf(paste0(fig_dir, 'metacell_new_old_module_overlap_xd.pdf'), width=6, height=5)
p
dev.off()






```


Try viveks loop for tree cutting

Not using

```{r eval=FALSE}

net <- GetNetworkData(seurat_obj)
geneTree <- net$dendrograms[[1]]
TOM <- GetTOM(seurat_obj)
dissTOM <- 1 - TOM
datExpr <- GetDatExpr(seurat_obj)

mColorh <- mLabelh <- colorLabels <- NULL
 for (minModSize in c(20,50,100)) {
   for (dthresh in c(0.05,0.1,0.2,0.25)) {
     for (ds in c(2,4)) {
       print("Trying parameters:")
       print(c(minModSize,dthresh,ds))
       tree = cutreeHybrid(
         dendro = geneTree,
         pamStage=FALSE,
         minClusterSize = minModSize,
         cutHeight = 0.9999,
         deepSplit = ds,
         distM = as.matrix(dissTOM)
       )
        merged <- mergeCloseModules(
          exprData = datExpr,
          colors = tree$labels,
          cutHeight = dthresh
        )
       mColorh <- cbind(mColorh,labels2colors(merged$colors))
       mLabelh <- c(mLabelh,paste("DS=",ds," mms=\n",minModSize," dcor=",dthresh))
     }
   }
 }

# mColorh1=cbind(mColorh,geneSigs[1,],geneSigs[2,],geneSigs[3,],geneSigs[4,],geneSigs[5,],geneSigs[6,])
# mLabelh1=c(mLabelh,rownames(geneSigs))

pdf(paste0(fig_dir, "SignedDendro2.pdf"),height=15,width=15)
plotDendroAndColors(
  geneTree,
  mColorh,
  groupLabels=mLabelh,
  addGuide=TRUE,
  dendroLabels=FALSE,
  main="Dendrogram With Different Module Cutting Parameters")
dev.off()



# update modules based on these parameters:

modules_orig <- GetModules(seurat_obj)
modules <- modules_orig
modules$module <- factor(as.character(mColorh[,12]))
modules$color <- as.character(mColorh[,12])

# update modules
seurat_obj <- SetModules(seurat_obj, modules)

colors <- as.character(modules$color)
names(colors) <- modules$gene_name
net$color <- colors

seurat_obj <- SetNetworkData(seurat_obj, net)

pdf(paste0(fig_dir,"dendro_update.pdf"),height=3, width=6)
PlotDendrogram(seurat_obj, main=paste0('hdWGCNA Dendrogram'))
dev.off()

table(GetModules(seurat_obj)$module)

```

Proceed with these modules 

```{r eval=FALSE}


# run RenameModules
seurat_obj <- ResetModuleNames(
  seurat_obj,
  new_name = "M"
)
print(names(GetModules(seurat_obj)))

library(MetBrewer)

modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct %>% arrange(module) %>% .$color
n_colors <- length(mod_colors) -1

new_colors <- paste0(met.brewer("Cross", n=n_colors, type='discrete'))
seurat_obj <- ResetModuleColors(seurat_obj, new_colors)




seurat_obj <- ScaleData(seurat_obj, features=rownames(seurat_obj)[1:100])
seurat_obj$Assignment <- droplevels(seurat_obj$Assignment)
seurat_obj <- ModuleEigengenes(
    seurat_obj,
    group.by.vars="Assignment" # snRNAseq batch
)

# compute module connectivity:
seurat_obj <- ModuleConnectivity(
  seurat_obj,
  group.by = 'cell_type', group_name = 'MHb-Neuron'
)


modules <- GetModules(seurat_obj)
subset(modules, gene_name == 'Nr4a2')
subset(modules, gene_name == 'Gabra2')
subset(modules, gene_name == 'Sox5')
table(modules$module)


plot_list <- ModuleFeaturePlot(
  seurat_obj, 
  order=TRUE, 
  raster=TRUE, 
  restrict_range=FALSE, 
  raster_dpi=400, 
  raster_scale=0.5, 
  point_size=1,
  reduction = 'neuron_umap'
  )

plot_list <- lapply(plot_list, function(x){
  x + NoLegend() + theme(plot.margin=margin(0,0,0,0))
})

pdf(paste0(fig_dir, "MHb_featureplot_MEs3.pdf"),height=6, width=12)
wrap_plots(plot_list, ncol=5)
dev.off()





# seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)
MEs <- GetMEs(seurat_obj)
mods <- levels(modules$module); mods <- mods[mods != 'grey']
seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)

p <- custom_vln(
    seurat_obj,
    features = mods,
    group.by = 'annotation',
    split.by = 'Group',
    groups = c('MHb-1', 'MHb-2', 'MHb-3', 'MHb-4', 'MHb-5'),
    add_boxplot=FALSE,
    split_colors=c('darkgoldenrod3', 'hotpink3'),
    add_colorbar=TRUE,
    plot_ymin = NA
  )

  #seurat_obj@meta.data <- seurat_obj@meta.data[,!(colnames(seurat_obj@meta.data) %in% colnames(MEs))]

pdf(paste0(fig_dir, 'MHb_hME_vln_stack.pdf'), width=3, height=5)
p
dev.off()





########################################
# Dot Plot
########################################


MEs <- GetMEs(seurat_obj)
modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mods <- mods[mods!='grey']

seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)


# make dotplot
p <- DotPlot(
  seurat_obj,
  group.by='annotation',
  features = rev(mods)
#  scale=FALSE, col.max=5
) + coord_flip() + RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue') + xlab('') + ylab('') +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=1)
  )

pdf(paste0(fig_dir, 'MHb_hMEs_dotplot2.pdf'), width=9, height=9)
p
dev.off()

# remove from seurat obj
seurat_obj@meta.data <- seurat_obj@meta.data[,!(colnames(seurat_obj@meta.data) %in% colnames(MEs))]




###################################################################################
# Run enrichr with these modules
##################################################################################


library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021','WikiPathway_2021_Human', 'KEGG_2021_Human')

# compute GO terms:
seurat_obj <- RunEnrichr(seurat_obj, dbs=dbs, max_genes=Inf)

# inspect the enrichr table:
enrichr_df <- GetEnrichrTable(seurat_obj) %>% subset(P.value < 0.05)

write.table(enrichr_df, quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'MHb_hdWGCNA_enrichr.tsv'))

# save the significant results
subset(enrichr_df, P.value < 0.05) %>% 
write.table(quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'MHb_hdWGCNA_enrichr_signif.tsv'))

# get module info and colors
modules <- GetModules(seurat_obj)
color_df <- modules %>% subset(module!='grey') %>%
  select(c(module, color)) %>% distinct %>%
  rename(c(group=module, colour=color))

# plot selected GO terms
combined_output <- GetEnrichrTable(seurat_obj)
combined_output $ngenes <- unlist(lapply(strsplit(combined_output $Genes, ';'), function(x){length(x)}))
combined_output <- subset(combined_output , ngenes >= 3)

selected_terms <- read.delim('data/MHb_hdWGCNA_enrichr_selected.txt', sep='\t', header=1)

# subset selected terms
combined_output$ngenes <- unlist(lapply(strsplit(combined_output$Genes, ';'), function(x){length(x)}))
selected_terms <- subset(combined_output, Term %in% selected_terms$Term & P.value < 0.05 & ngenes > 2)


selected_terms$group <- factor(
  as.character(selected_terms$module),
  levels = mods
)

# set max pval

quantile(-log(selected_terms$P.value), 0.95)
max_p <- 10

selected_terms$logp <- -log(selected_terms$P.value)
selected_terms$logp <- ifelse(selected_terms$logp > max_p, max_p, selected_terms$logp)

# remove GO Term ID
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")

selected_terms <- selected_terms %>%
  arrange(group)


#selected_terms$wrap <- wrapText(selected_terms$Term, 35)

selected_terms$Term <- factor(
  as.character(selected_terms$Term),
  levels = rev(unique(as.character(selected_terms$Term)))
)

# GO Term dot plot

p <- selected_terms %>%
  ggplot(aes(x = group, y = Term, color =logp, size=log(Combined.Score))) +
  geom_point() +
  scale_color_stepsn(colors=rev(magma(256))) +
  RotatedAxis() + xlab('') + ylab('') +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.margin = margin(0,0,0,0),
    panel.grid = element_line(size=0.25, color='lightgrey')
  )+ labs(
    color = bquote("-log"[10]~"(P)"),
    size= bquote("log"[10]~"(Enrich)")
  )


# # make the colorbar as its own heatmap
color_df$var <- 1
cp <- color_df$colour; names(cp) <- color_df$group
colorbar <- color_df %>%
  subset(group %in% unique(selected_terms$group)) %>%
  ggplot(aes(x=group, y=var, fill=group)) +
  geom_tile() +
  scale_fill_manual(values=cp) +
  coord_equal() +
  NoLegend() + RotatedAxis() +
  theme(
    plot.title=element_blank(),
    axis.line=element_blank(),
    axis.ticks.y =element_blank(),
    axis.text.y = element_blank(),
    axis.title = element_blank(),
    plot.margin=margin(0,0,0,0),
  )




pdf(paste0(fig_dir, 'selected_go_terms_bubbleplot.pdf'), width=7.3, height=7)
p / colorbar
dev.off()


################################################################################
# Hubgene circle plots:
################################################################################

library(igraph)

# individual module networks
ModuleNetworkPlot(
  seurat_obj,
  mods = "all",
  #label_center=TRUE,
  outdir = paste0(fig_dir, 'MHb_hubNetworks2/')
)


################################################################################
# UMAP:
################################################################################


seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 3,
  n_neighbors=10,
  min_dist=0.3,
  spread=3,
  supervised=TRUE,
  target_weight=0.5
)


# get the hub gene UMAP table from the seurat object
umap_df <- GetModuleUMAP(seurat_obj)

# plot with ggplot
p <- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(
   color=umap_df$color,
   size=umap_df$kME*2
  ) +
  umap_theme()

pdf(paste0(fig_dir, 'MHb_hubgene_umap_ggplot2.pdf'), width=5, height=5)
p
dev.off()


label_genes <- c('Gabra2', 'Nr4a2')
library(reshape2)
library(igraph)
pdf(paste0(fig_dir, 'MHb_hubgene_umap_igraph.pdf'), width=10, height=10)
ModuleUMAPPlot(
  seurat_obj,
  edge.alpha=0.5,
  sample_edges=TRUE,
  keep_grey_edges=FALSE,
  edge_prop=0.075, # taking the top 20% strongest edges in each module
  label_hubs=3, # how many hub genes to plot per module?
  label_genes = label_genes
)
dev.off()


```


DME analysis to compare Nurr2c vs GFP 

and marker DME analysis  

```{r eval=FALSE}

# marker DMEs 
dmes <- FindAllDMEs(
  seurat_obj,
  group.by = 'cell_type'
)
subset(dmes, group == 'MHb-Neuron' & avg_log2FC > 0)


groups <- as.character(unique(seurat_obj$annotation))

DMEs <- data.frame()
for(cur_group in groups){

  group1 <- seurat_obj@meta.data %>% subset(annotation == cur_group & Group == "Nurr2c") %>% rownames
  group2 <- seurat_obj@meta.data %>% subset(annotation == cur_group & Group == "GFP") %>% rownames

  head(group1)

  cur_DMEs <- FindDMEs(
    seurat_obj,
    barcodes1 = group1,
    barcodes2 = group2,
    test.use='wilcox',
    pseudocount.use=0.01
  )
  cur_DMEs$group <- cur_group 
  DMEs <- rbind(DMEs, cur_DMEs)

}

subset(DMEs, group == 'MHb-1')
subset(DMEs, module == 'M1')




maxval <- 0.5; minval <- -0.5
plot_df <- DMEs
plot_df$avg_log2FC <- ifelse(plot_df$avg_log2FC > maxval, maxval, plot_df$avg_log2FC)
plot_df$avg_log2FC <- ifelse(plot_df$avg_log2FC < minval, minval, plot_df$avg_log2FC)

plot_df$textcolor <- ifelse(plot_df$avg_log2FC > 0.2, 'black', 'black')
plot_df$Significance <- gtools::stars.pval(plot_df$p_val_adj)

# set factor levels:
#plot_df$module <- factor(as.character(plot_df$module), levels=mods)

p <- plot_df %>% 
  ggplot(aes(x=group, y=fct_rev(module), fill=avg_log2FC)) +
  geom_tile() +
  geom_text(label=plot_df$Significance, color=plot_df$textcolor) +
  scale_fill_gradient2(low='hotpink3', mid='grey95', high='goldenrod3') +
  RotatedAxis() +
 # labs(fill = 'Odds ratio') +
  theme(
    panel.border = element_rect(fill=NA, color='black', size=1),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    #axis.title.x = element_blank(),
    #axis.title.y = element_blank(),
    #axis.text.x = element_blank(),
   # axis.ticks.x = element_blank(),
   # axis.text.y = element_blank(),
    #axis.ticks.y = element_blank(),
    plot.margin=margin(0,0,0,0)
  ) + xlab('') + ylab('') 

# Plot the result
pdf(paste0(fig_dir, 'DME_heatmap2.pdf'),height=2.5, width=4)
p 
dev.off()



```


Saving 

```{r eval=FALSE}

saveRDS(seurat_obj, file = paste0(data_dir, 'harmony_annotated_hdWGCNA.rds'))

```


Overlap of Nr4a2 target genes and modules 

```{r eval=FALSE}

#---------------------------------------------------------------------------#
# Make a table of primary and secondary Nr4a2 target genes 
#---------------------------------------------------------------------------#

cur_tf <- 'Nr4a2'

tf_nets <- dir('../tf_net/data/tf_nets/')
tf_nets <- tf_nets[grepl('importance', tf_nets)]


# parameters for regulons
n_tfs <- 5
importance_thresh <- 0.001
combined_output <- data.frame()

primary_genes <- list()
secondary_genes <- list()

for(cur_net_file in tf_nets){

  print(cur_net_file)

  # load the tf-gene table
  importance_df <- read.csv(paste0('../tf_net/data/tf_nets/', cur_net_file ))

  tmp <- strsplit(cur_net_file, '_')[[1]]
  cur_celltype <- tmp[2]
  cur_group <- tmp[3]

  #---------------------------------------------------------------------------#
  # Define the TF regulons
  #---------------------------------------------------------------------------#

  regulons <- importance_df %>% 
    subset(Gain > importance_thresh) %>% 
    group_by(gene) %>%
    slice_max(order_by=Gain, n=n_tfs) %>% 
    ungroup()

  # compute the degree for each TF:
  tf_degrees <- table(regulons$tf)

  #---------------------------------------------------------------------------#
  # Get the primary & secondary targets of Nr4a2
  #---------------------------------------------------------------------------#

  # primary target genes 
  cur_primary<- regulons %>% 
    subset(tf == cur_tf) 

  # which of these pimary target genes are tfs?
  cur_primary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  cur_tfs <- unique(c(cur_tf, cur_primary_tfs))

  # get the regulons for these TFs:
  cur_secondary <- subset(regulons, tf %in% cur_primary_tfs)
  cur_secondary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  primary_genes[[paste0(cur_celltype, '_', cur_group)]] <- cur_primary$gene 
  secondary_genes[[paste0(cur_celltype, '_', cur_group)]] <- cur_secondary$gene

}

# tf_targets <- list(
#     'secondary' = 
primary <- primary_genes[['MHb-Neuron_Nurr2c']]
secondary <- secondary_genes[['MHb-Neuron_Nurr2c']]
secondary <- setdiff(secondary, primary)


modules <- GetModules(seurat_obj)
mods <- levels(modules$module); mods <- mods[mods != 'grey']


genome.size <- nrow(seurat_obj)

overlap_df <- data.frame()
plot_df <- data.frame()
for(cur_mod in mods){

  cur_genes <- subset(modules, module == cur_mod) %>% .$gene_name

  cur_primary <- intersect(cur_genes, primary)
  cur_secondary <- intersect(cur_genes, secondary)
  cur_other <- setdiff(cur_genes, unique(c(cur_primary, cur_secondary)))

  cur_df <- data.frame(
    module = cur_mod,
    target_type = c('primary', 'secondary', 'other'),
    n = c(
      length(cur_primary),
      length(cur_secondary),
      length(cur_other)
    )
  )
  cur_df$percentage <- cur_df$n / sum(cur_df$n)
  plot_df <- rbind(plot_df, cur_df)

  cur_overlap_primary <- testGeneOverlap(newGeneOverlap(
      primary,
      cur_genes,
      genome.size=genome.size
  ))
  cur_overlap_secondary <- testGeneOverlap(newGeneOverlap(
      secondary,
      cur_genes,
      genome.size=genome.size
  ))

  cur_overlap <- data.frame(
    'odds.ratio' = c(cur_overlap_primary@odds.ratio, cur_overlap_secondary@odds.ratio),
    'pval' = c(cur_overlap_primary@pval, cur_overlap_secondary@pval),
    'Jaccard' = c(cur_overlap_primary@Jaccard, cur_overlap_secondary@Jaccard),
    'size_intersection' = c(length(cur_overlap_primary@intersection), length(cur_overlap_secondary@intersection)),
    'target_type' = c('primary', 'secondary'),
    'module' = cur_mod
  )
  overlap_df <- rbind(overlap_df, cur_overlap)

}
overlap_df$fdr <- p.adjust(overlap_df$pval, method='fdr')
overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)

# plot the stacked bar chart 

plot_df$target_type <- factor(as.character(plot_df$target_type), levels=c('other', 'secondary', 'primary'))

p1 <- plot_df %>%
  ggplot(aes(x = percentage, y=fct_rev(module), fill=target_type)) + 
  geom_bar(position='stack', stat='identity') + 
  # scale_fill_manual(values=amylo_cp) + 
  geom_text(aes(label=abs(n)), position = position_stack(vjust=0.5)) +
  #xlim(-plot_max, plot_max) +
#  scale_x_continuous(expand = c(0, 0), limits = c(-plot_max, plot_max)) + 
  theme(
    axis.line.y = element_blank(),
    axis.title.y = element_blank()
  ) + 
  xlab("Proportion") + NoLegend() + 
   scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) 
  #facet_wrap(~direction, ncol=2) + NoLegend()
#      ylab(bquote("-log"[10]~"(Adj. P-value)")) +

pdf(paste0(fig_dir, 'Nr4a2_target_modules_barplot.pdf'), width=3, height=4.5, useDingbats=FALSE)
p1
dev.off()

library(ggbreak)

p2 <- overlap_df %>% 
  ggplot(aes(x=fct_rev(module), y = odds.ratio, size=-log10(fdr), color=target_type, fill=target_type)) + 
  geom_hline(yintercept=1, color='lightgrey', linetype='dashed') +
     geom_linerange(aes(x=fct_rev(module), xmax=fct_rev(module), ymin=0, ymax=odds.ratio), size=0.5, position=position_dodge(width=1)) +
  geom_point(position=position_dodge(width=1), shape=overlap_df$shape, color='black') +
  coord_flip() +
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA)) + ylab('') + xlab('')

pdf(paste0(fig_dir, 'Nr4a2_target_modules_lollipop_legend.pdf'), width=4, height=4.5, useDingbats=FALSE)
p2
dev.off()


```

Overlap of DEGs and modules 

```{r eval=FALSE}

# celltype DEGs
degs <- read.csv(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/revision/DEGs/data/celltype_Nurr2c_vs_GFP_DEGs.csv')
#degs <- subset(degs, group %in% seurat_obj$cell_type)
degs$group <- factor(
  degs$group,
  levels = levels(seurat_obj$cell_type)
)
name <- 'celltype'
color1 <- 'darkgoldenrod3'; color2 <- 'hotpink3'

cur_degs <- subset(degs, group == 'MHb-Neuron' & p_val_adj < 0.05)
cur_degs[grepl('Srsf', cur_degs$gene),]

modules <- GetModules(seurat_obj)
cur_genes <- subset(modules, module == 'M2') %>% .$gene_name
cur_genes[grepl('Srsf', cur_genes)]
cur_genes[grepl('', cur_genes)]


cur_group <- 'MHb-Neuron'
fc_cutoff <- 0.25
cur_up <- degs %>% subset(group == cur_group & p_val_adj < 0.05 & avg_log2FC >= fc_cutoff) %>% .$gene
cur_down <- degs %>% subset(group == cur_group & p_val_adj < 0.05 & avg_log2FC <= -1* fc_cutoff) %>% .$gene
#up_setsize <- length(intersect)

genome.size <- nrow(seurat_obj)
overlap_df <- data.frame()
plot_df <- data.frame()
for(cur_mod in mods){

  cur_genes <- subset(modules, module == cur_mod) %>% .$gene_name

  cur_overlap_up <- testGeneOverlap(newGeneOverlap(
      cur_up,
      cur_genes,
      genome.size=genome.size
  ))
  cur_overlap_down <- testGeneOverlap(newGeneOverlap(
      cur_down,
      cur_genes,
      genome.size=genome.size
  ))

  cur_overlap <- data.frame(
    'odds.ratio' = c(cur_overlap_up@odds.ratio, cur_overlap_down@odds.ratio),
    'pval' = c(cur_overlap_up@pval, cur_overlap_down@pval),
    'Jaccard' = c(cur_overlap_up@Jaccard, cur_overlap_down@Jaccard),
    'size_intersection' = c(length(cur_overlap_up@intersection), length(cur_overlap_down@intersection)),
    'direction' = c('up', 'down'),
    'module' = cur_mod
  )
  overlap_df <- rbind(overlap_df, cur_overlap)

}
overlap_df$fdr <- p.adjust(overlap_df$pval, method='fdr')
overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)


```


Assign all TFs to an actual module rather than a co-expression module 

```{r eval=FALSE}

modules <- GetModules(seurat_obj, wgcna_name)
mod_colors <- dplyr::select(modules, c(module, color)) %>% distinct()
mod_cp <- mod_colors$color; names(mod_cp) <- as.character(mod_colors$module)
mods <- levels(modules$module); mods <- mods[mods != 'grey']


grey_genes <- subset(modules, module == 'grey') %>% .$gene_name

# get a list of all TFs:
importance_df <- read.csv(paste0('../tf_net/data/tf_nets/TFnet_MHb-Neuron_Nurr2c_importance.csv'))
tf_list <- unique(importance_df$tf)

grey_tfs <- tf_list[tf_list %in% grey_genes]

# get just the kME values from the modules table
kMEs <- modules[,4:ncol(modules)]
kMEs <- kMEs[,colnames(kMEs) != "kME_grey"]


reassigned <- sapply(grey_tfs, function(cur_gene){
  cur_kMEs <- kMEs[cur_gene,]
  max_kME <- max(cur_kMEs)
  if(max_kME < 0){
    return('kME_grey')
  }
  colnames(kMEs)[which(cur_kMEs == max_kME)]
})


# add the reassigned modules to the modules table
reassigned <- do.call(rbind, strsplit(reassigned, 'kME_'))[,2]
reassigned <- factor(as.character(reassigned), levels=levels(modules$module))

# new colors:
reassigned_colors <- as.character(mod_cp[as.character(reassigned)])

# reassign modules and colors
modules[grey_tfs,'module'] <- reassigned
modules[grey_tfs,'color'] <- reassigned_colors

subset(modules, gene_name %in% grey_tfs) %>% .$module %>% table

seurat_obj <- SetModules(seurat_obj, modules)


```

Integration / visualization of the Nr4a2 target genes with the TF reg net 

```{r eval=FALSE}

modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct %>% arrange(module) %>% .$color
cp <- mod_colors; names(cp) <- mods
mods <- mods[mods != 'grey']

# get a list of all TFs:
importance_df <- read.csv(paste0('../tf_net/data/tf_nets/TFnet_MHb-Neuron_Nurr2c_importance.csv'))
tf_list <- unique(importance_df$tf)

cur_tf <- 'Nr4a2'

tf_nets <- dir('../tf_net/data/tf_nets/')
tf_nets <- tf_nets[grepl('importance', tf_nets)]
tf_nets <- c('TFnet_MHb-Neuron_Nurr2c_importance.csv')
n_tfs <- 10
importance_thresh <- 0.001
network_df <- data.frame()

for(cur_net_file in tf_nets){

  print(cur_net_file)

  tmp <- strsplit(cur_net_file, '_')[[1]]
  cur_celltype <- tmp[2]
  cur_group <- tmp[3]

  importance_df <- read.csv(paste0('../tf_net/data/tf_nets/', cur_net_file ))

  #---------------------------------------------------------------------------#
  # Define the TF regulons
  #---------------------------------------------------------------------------#

  regulons <- importance_df %>% 
    subset(Gain > importance_thresh) %>% 
    group_by(gene) %>%
    slice_max(order_by=Gain, n=n_tfs) %>% 
    ungroup()

  # compute the degree for each TF:
  tf_degrees <- table(regulons$tf)

  #---------------------------------------------------------------------------#
  # Get the primary & secondary targets of Nr4a2
  #---------------------------------------------------------------------------#

  # primary target genes 
  cur_primary<- regulons %>% 
    subset(tf == cur_tf) 

  # which of these pimary target genes are tfs?
  cur_primary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  cur_tfs <- unique(c(cur_tf, cur_primary_tfs))

  # get the regulons for these TFs:
  cur_secondary <- subset(regulons, tf %in% cur_primary_tfs)
  cur_secondary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  # combine the primary and secondary into one table 
  cur_network <- rbind(cur_primary, cur_secondary)
  cur_network$Gain <- cur_network$Gain * sign(cur_network$Cor)

  cur_genes <- unique(cur_network$gene)
  length(cur_genes)

  # make an igraph network from the nr4a2 regulon:
  cur_network <- cur_network %>%
    dplyr::rename(c(source=tf, target=gene)) %>%
    mutate(Score = sign(Cor) * Gain)

  primary_genes <- unique(cur_primary$gene)
  secondary_genes <- unique(cur_network$target[! cur_network$target %in% primary_genes])

  cur_network$target_type <- ifelse(cur_network$target %in% primary_genes, 'primary', 'secondary')
  cur_network$cell_group <- cur_celltype 
  cur_network$group <- cur_group

  network_df <- rbind(network_df, cur_network)
}


# get a list of co-expression modules that 

modules <- GetModules(seurat_obj)
module_tfs <- subset(modules, gene_name %in% tf_list)


length(unique(network_df$source))

umap_df <- GetModuleUMAP(seurat_obj)

cur_network <- subset(network_df, group == 'Nurr2c' & cell_group == 'MHb-Neuron')
cur_network <- subset(cur_network, source %in% module_tfs$gene_name & target %in% module_tfs$gene_name)
cur_tfs <- unique(c(cur_network$source, cur_network$target))
cur_tfs <- cur_tfs[cur_tfs %in% modules$gene_name]




#---------------------------------------------------------------------------#
# Plot with ggraph
#---------------------------------------------------------------------------#

library(ggraph)
library(tidygraph)

graph <- as_tbl_graph(cur_network) %>% 
  activate(nodes) %>% 
  mutate(degree  = centrality_degree())  

tmp <- tf_degrees[names(V(graph))]; tmp <- tmp[!is.na(tmp)]
V(graph)[names(tmp)]$degree <- as.numeric(tmp)

V(graph)$gene_type <- ifelse(names(V(graph)) %in% unique(regulons$tf), 'TF', 'Gene')
V(graph)$gene_type <- ifelse(names(V(graph)) == cur_tf, cur_tf, V(graph)$gene_type)

# make the layout table using the umap coords:
umap_layout <- umap_df[names(V(graph)),] %>% dplyr::rename(c(x=UMAP1, y = UMAP2, name=gene))
rownames(umap_layout) <- 1:nrow(umap_layout)
lay <- create_layout(graph, umap_layout)

# set.seed(12345)
# n_pivots <- 250
# if(length(V(graph)) < n_pivots){
#   n_pivots <- length(V(graph)) / 2
# }
# lay <- create_layout(graph, layout='sparse_stress', pivots=n_pivots)

#lay <- create_layout(graph, layout='focus', focus = node_is_center())

# add extra info
# lay$tf_name <- ifelse(lay$name %in% unique(regulons$tf), lay$name, NA)
lay$tf_name <- ifelse(lay$name %in% c(cur_tf, cur_primary_tfs), lay$name, NA)
lay$size <- ifelse(lay$name %in% unique(regulons$tf), 5, 2)
lay$type <- ifelse(lay$name %in% primary_genes, 'Primary', 'Secondary')
lay$type <- ifelse(lay$name == cur_tf, cur_tf, lay$type)
lay$type <- factor(lay$type, levels = c(cur_tf, 'Primary', 'Secondary'))

# shape layout:
cur_shapes <- c(18, 17, 16); names(cur_shapes) <- c(cur_tf, 'TF', 'Gene')
cur_shapes <- c(23, 24, 25); names(cur_shapes) <- c(cur_tf, 'Primary', 'Secondary')

p <- ggraph(lay) + 
  ggrastr::rasterise(
    geom_point(inherit.aes=FALSE, data=umap_df, aes(x=UMAP1, y=UMAP2), color=umap_df$color, alpha=0.1, size=3),
    dpi=500
   ) +
  geom_edge_fan(
    aes(color=Cor, alpha=abs(Cor)),
    arrow = arrow(length = unit(2, 'mm'), type='closed'), 
    end_cap = circle(3, 'mm')
  ) + 
  geom_node_point(data=subset(lay, (! name %in% regulons$tf) | name == cur_tf ), aes(fill=module, shape=type, size=degree), color='black') +
  geom_node_point(data=subset(lay, name %in% regulons$tf & name != cur_tf), aes(fill=module, size=degree, shape=type), color='black') +

# ggrastr::rasterise(geom_node_point(aes(color=type, size=degree)), dpi=200) +
  #geom_node_text(aes(label=tf_name), repel=TRUE, max.overlaps=Inf, fontface='italic') +
  geom_node_label(aes(label=tf_name, color=module), repel=TRUE, max.overlaps=Inf, fontface='italic') +
  scale_edge_colour_gradient2(high='orange2', mid='white', low='dodgerblue')  + 
  scale_colour_manual(values=cp) + 
  scale_fill_manual(values=cp) + 
  scale_shape_manual(values=cur_shapes) 

pdf(paste0(fig_dir,'test_network4.pdf'), width=8, height=7)
print(p)
dev.off()

#---------------------------------------------------------------------------#
# Quantify the number of TF links between modules and plot as a heamap
#---------------------------------------------------------------------------#

tf_regulons <- subset(regulons, tf %in% module_tfs$gene_name & gene %in% module_tfs$gene_name)
ix <- match(tf_regulons$tf, modules$gene_name)
tf_regulons$source_module <- modules$module[ix]
ix <- match(tf_regulons$gene, modules$gene_name)
tf_regulons$target_module <- modules$module[ix]
tf_regulons %<>% dplyr::rename(source=tf, target=gene) %>% mutate(Gain = Gain * sign(Cor))
cur_network <- tf_regulons

# add module info to the network:
ix <- match(cur_network$source, modules$gene_name)
cur_network$source_module <- modules$module[ix]
ix <- match(cur_network$target, modules$gene_name)
cur_network$target_module <- modules$module[ix]

# make empty matrices to store the number of links between mods
pos_mat <- matrix(0, length(mods), length(mods))
rownames(pos_mat) <- mods; colnames(pos_mat) <- mods
neg_mat <- matrix(0, length(mods), length(mods))
rownames(neg_mat) <- mods; colnames(neg_mat) <- mods

# loop through combos of mods and identify the number of +/- links between mods
combos <- expand.grid(mods, mods)
for(i in 1:nrow(combos)){
  m1 <- as.character(combos[i,'Var1'])
  m2 <- as.character(combos[i, 'Var2'])

  # how many positive connections from m1 to m2:
  cur_pos <- subset(cur_network, target_module == m1 & source_module == m2 & Gain >= 0)
  cur_neg <- subset(cur_network, target_module == m1 & source_module == m2 & Gain < 0)

  if(m1 == m2){
    pos_mat[m1,m2] <- 0
    neg_mat[m1,m2] <- 0
  } else{
    pos_mat[m1,m2] <- nrow(cur_pos)
    neg_mat[m1,m2] <- nrow(cur_neg)
  }

}

max_val <- 1
plot_df <- reshape2::melt(neg_mat)
plot_df$count <-plot_df$value
plot_df$label <- ifelse(plot_df$value > 5, plot_df$value, "")
tmp <-table(module_tfs$module)
ix <- match(plot_df$Var1, names(tmp))
plot_df$value <- plot_df$value / as.numeric(tmp)[ix]
plot_df1 <- plot_df

plot_df$value <- ifelse(plot_df$value > max_val, max_val, plot_df$value)
p1 <- plot_df %>% 
  ggplot(aes(x=Var1, y=fct_rev(Var2), fill=value)) + 
  geom_tile() + 
  geom_text(aes(label=label)) +
  scale_fill_gradient(low='grey95', high='dodgerblue') + 
  coord_fixed() + RotatedAxis() + 
  xlab('Target') + ylab('Source') + 
  ggtitle('Repressive interactions') + 
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(linewidth=1, fill=NA, color='black')
  )


plot_df <- reshape2::melt(pos_mat)
plot_df$count <-plot_df$value
plot_df$label <- ifelse(plot_df$value > 5, plot_df$value, "")

tmp <-table(module_tfs$module)
ix <- match(plot_df$Var1, names(tmp))
plot_df$value <- plot_df$value / as.numeric(tmp)[ix]
plot_df2 <- plot_df

plot_df$value <- ifelse(plot_df$value > max_val, max_val, plot_df$value)
p2 <- plot_df %>% 
  ggplot(aes(x=Var1, y=fct_rev(Var2), fill=value)) + 
  geom_tile() + 
  geom_text(aes(label=label)) +
  scale_fill_gradient(low='grey95', high='orange2') + 
  coord_fixed() + RotatedAxis()  + 
  xlab('Target') + ylab('Source') + 
  ggtitle('Activating interactions') + 
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(linewidth=1, fill=NA, color='black')
  )


pdf(paste0(fig_dir,'test_network_interactions_heatmap_norm.pdf'), width=7, height=4)
p1 + p2 + plot_layout(guides='collect')
dev.off()

#---------------------------------------------------------------------------#
# Delta of Activating / Repressive interactions
#---------------------------------------------------------------------------#

plot_df1$delta <-  plot_df2$value - plot_df1$value

p3 <- plot_df1 %>% 
  ggplot(aes(x=Var1, y=fct_rev(Var2), fill=delta)) + 
  geom_tile() + 
  #geom_text(aes(label=label)) +
  scale_fill_gradient2(low='dodgerblue', mid='grey95', high='orange2') + 
  coord_fixed() + RotatedAxis()  + 
  xlab('Target') + ylab('Source') + 
  ggtitle('Activating - repressing') + 
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(linewidth=1, fill=NA, color='black')
  )


pdf(paste0(fig_dir,'test_network_interactions_heatmap_norm_delta.pdf'), width=4, height=10)
p1 / p2 / p3 
dev.off()

#---------------------------------------------------------------------------#
# Dot Plot to show expression of these TFs
#---------------------------------------------------------------------------#

plot_genes <- subset(cur_network, target_type == 'primary') %>% .$target %>% unique
plot_genes <- c('Nr4a2',plot_genes)
plot_genes <- modules %>% arrange(module) %>% subset(gene_name %in% plot_genes)

# make dotplot
p <- DotPlot(
  subset(seurat_obj, Group == 'Nurr2c'),
  group.by='annotation',
  features = rev(plot_genes$gene_name)
#  scale=FALSE, col.max=5
) + coord_flip() + RotatedAxis() +
  scale_color_gradient(high='darkorchid3', low='lightgrey') + xlab('') + ylab('') +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=1),
    axis.text.y = element_text(color=rev(plot_genes$color), face='italic')
  )

pdf(paste0(fig_dir, 'tf_expression_dotplot.pdf'), width=4.25, height=6)
p
dev.off()

#---------------------------------------------------------------------------#
# Bar plot showing the top and bottom TF targets of nr4a2 
#---------------------------------------------------------------------------#

plot_genes <- subset(cur_network, target_type == 'primary') %>% .$target %>% unique
plot_genes <- c('Nr4a2',plot_genes)
plot_genes <- modules %>% arrange(module) %>% subset(gene_name %in% plot_genes)


plot_df <- subset(cur_network, source == 'Nr4a2' & target %in% plot_genes$gene_name)
plot_df %<>% mutate(rank = dense_rank(dplyr::desc(Gain))) %>% 
  arrange(Gain, desc=TRUE)


p <- plot_df %>%
  ggplot(aes(y=rev(factor(rank)), x = Gain, fill=Gain)) + 
  geom_bar(stat='identity', width=1) +
  geom_vline(xintercept=0, color='black') + 
  geom_text(aes(label=target), color='black', size=3.5, hjust='center') +
  scale_fill_gradient2(high="orange2", mid='grey95', low='dodgerblue') +
  NoLegend() + xlab('') + ylab('') +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.line.y = element_blank()
  )

pdf(paste0(fig_dir, 'Nr4a2_target_ranks.pdf'), width=3, height=4)
p
dev.off()


```


Module Preservation analysis with the Hashikawa dataset 
and the Naive dataset

```{r eval=FALSE}

# load the hashikawa dataset 
seurat_hashikawa <- readRDS('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/analysis/test_harmony/data/hashikawa_seurat.rds')
seurat_hashikawa$cell_type <- ifelse(grepl("MHb", seurat_hashikawa$celltype_neuron), 'MHb', seurat_hashikawa$celltype_neuron)

seurat_hashikawa <- ProjectModules(
  seurat_obj = seurat_hashikawa,
  seurat_ref = seurat_obj,
  wgcna_name_proj="MHb_projected",
  wgcna_name = "MHb_metacell"
)


# plot modules in hashikawa data:
plot_list <- ModuleFeaturePlot(
  seurat_hashikawa, order="shuffle", raster=TRUE,
  restrict_range=FALSE, point_size=0.5, raster_scale=0.5
)

# remove legend, title, then plot
for(i in 1:length(plot_list)){
  plot_list[[i]] <- plot_list[[i]] + NoLegend() + ggtitle('') + theme(plot.margin = margin(0,0,0,0))
}

pdf("figures/MHb_featureplot_hMEs_hashikawa.pdf",height=6, width=12)
wrap_plots(plot_list, ncol=5)
dev.off()



################################################################################
# Module preservation
################################################################################


seurat_hashikawa <- SetDatExpr(
  seurat_hashikawa,
  group_name = "MHb",
  group.by = "cell_type",
  use_metacells = FALSE
)

# run module preservation function
seurat_hashikawa <- ModulePreservation(
  seurat_hashikawa,
  seurat_ref = seurat_obj,
  name="MHb_projected",
  verbose=3,
  n_permutations=200
)

plot_list <- PlotModulePreservation(
  seurat_hashikawa,
  name="MHb_projected",
  statistics = "summary"
)

pdf(paste0(fig_dir, 'hashikawa_module_preservation_summary.pdf'), width=10, height=5)
wrap_plots(plot_list, ncol=2)
dev.off()


saveRDS(seurat_hashikawa, file='data/hashikawa_hdWGCNA.rds')
seurat_hashikawa <- readRDS('data/hashikawa_hdWGCNA.rds')

################################################################################
# Preservation in the naive dataset:
################################################################################


seurat_full <- readRDS(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/revision/data/harmony_annotated_integration.rds')
seurat_naive <- subset(seurat_full, cell_type == 'MHb-Neuron' & Group %in% c('NN', 'NGFP'))


seurat_naive <- ProjectModules(
  seurat_obj = seurat_naive,
  seurat_ref = seurat_obj,
  wgcna_name_proj="MHb_projected",
  wgcna_name = "MHb_metacell"
)



################################################################################
# Module preservation
################################################################################


seurat_naive <- SetDatExpr(
  seurat_naive,
  group_name = "MHb-Neuron",
  group.by = "cell_type",
  use_metacells = FALSE
)

# run module preservation function
seurat_naive <- ModulePreservation(
  seurat_naive,
  seurat_ref = seurat_obj,
  name="MHb_projected",
  verbose=3,
  n_permutations=200
)

plot_list <- PlotModulePreservation(
  seurat_naive,
  name="MHb_projected",
  statistics = "summary"
)

pdf(paste0(fig_dir, 'naive_module_preservation_summary.pdf'), width=10, height=5)
wrap_plots(plot_list, ncol=2)
dev.off()


saveRDS(seurat_naive, file='data/naive_hdWGCNA.rds')
seurat_naive <- readRDS(file='data/naive_hdWGCNA.rds')

```











GSEA of co-ex modules and Nr4a2 TF target genes
Note: I don't think we should use this result

```{r eval=FALSE}

#---------------------------------------------------------------------------#
# Make a table of primary and secondary Nr4a2 target genes 
#---------------------------------------------------------------------------#

cur_tf <- 'Nr4a2'

tf_nets <- dir('../tf_net/data/tf_nets/')
tf_nets <- tf_nets[grepl('importance', tf_nets)]


# parameters for regulons
n_tfs <- 5
importance_thresh <- 0.001
combined_output <- data.frame()

primary_genes <- list()
secondary_genes <- list()

for(cur_net_file in tf_nets){

  print(cur_net_file)

  # load the tf-gene table
  importance_df <- read.csv(paste0('../tf_net/data/tf_nets/', cur_net_file ))

  tmp <- strsplit(cur_net_file, '_')[[1]]
  cur_celltype <- tmp[2]
  cur_group <- tmp[3]

  #---------------------------------------------------------------------------#
  # Define the TF regulons
  #---------------------------------------------------------------------------#

  regulons <- importance_df %>% 
    subset(Gain > importance_thresh) %>% 
    group_by(gene) %>%
    slice_max(order_by=Gain, n=n_tfs) %>% 
    ungroup()

  # compute the degree for each TF:
  tf_degrees <- table(regulons$tf)

  #---------------------------------------------------------------------------#
  # Get the primary & secondary targets of Nr4a2
  #---------------------------------------------------------------------------#

  # primary target genes 
  cur_primary<- regulons %>% 
    subset(tf == cur_tf) 

  # which of these pimary target genes are tfs?
  cur_primary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  cur_tfs <- unique(c(cur_tf, cur_primary_tfs))

  # get the regulons for these TFs:
  cur_secondary <- subset(regulons, tf %in% cur_primary_tfs)
  cur_secondary_tfs <- cur_primary %>% 
    subset(gene %in% unique(regulons$tf)) %>% .$gene

  primary_genes[[paste0(cur_celltype, '_', cur_group)]] <- cur_primary$gene 
  secondary_genes[[paste0(cur_celltype, '_', cur_group)]] <- cur_secondary$gene

}

#---------------------------------------------------------------------------#
# GSEA for each module 
#---------------------------------------------------------------------------#

modules <- GetModules(seurat_obj)
hub_df <- GetHubGenes(seurat_obj, n_hubs=Inf)

# add ranks to this table:

hub_df %<>% group_by(module) %>% mutate(rank = dense_rank(desc(kME)))

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes_full <- hg38_mm10_genes
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))
hg38_mm10_genes <- subset(hg38_mm10_genes, mm10_name != '' & hg38_name != '')

# need to make sure that there's only one entry for each gene in hg38_mm10_genes
mm10_genes <- unique(hg38_mm10_genes$mm10_name)
hg38_genes <- unique(hg38_mm10_genes$hg38_name)
hg38_mm10_genes <- hg38_mm10_genes[match(mm10_genes, hg38_mm10_genes$mm10_name),]

ix <- match(hub_df$gene_name, hg38_mm10_genes$mm10_name)
hub_df$human_gene <- hg38_mm10_genes[ix,'hg38_name']
hub_df$human_gene <- ifelse(is.na(hub_df$human_gene), toupper(hub_df$gene_name), hub_df$human_gene)













modules <- GetModules(seurat_obj)
hub_df <- GetHubGenes(seurat_obj, n_hubs=Inf)

# add ranks to this table:

hub_df %<>% group_by(module) %>% mutate(rank = dense_rank(desc(kME)))

gsea_df <- data.frame()
for(cur_module in mods){

  print(cur_module)

  # weight for current module:
  cur_hubs <- subset(hub_df, module == cur_module)

  ranks <- cur_hubs$kME; names(ranks) <- cur_hubs$gene_name
  ranks <- ranks[order(ranks)]

  tf_targets <- list(
    'primary' = subset(network_df, source == 'Nr4a2' & target_type == 'primary' & cell_group == 'MHb-Neuron' & group == 'Nurr2c') %>% .$target,
    'secondary' = subset(network_df, source == 'Nr4a2' & target_type == 'secondary' & cell_group == 'MHb-Neuron' & group == 'Nurr2c') %>% .$target
  )

  tf_targets <- list(
    'primary' = primary_genes[['MHb-Neuron_Nurr2c']],
    'secondary' = secondary_genes[['MHb-Neuron_Nurr2c']]
  )

  # run fgsea to compute enrichments
  fgseaRes <- fgsea(
    #pathways = pathways, 
    pathways = tf_targets,
    stats    = ranks,
    minSize  = 15,
    maxSize  = Inf,
    scoreType = 'pos'
  )
  fgseaRes$module <- cur_module
  gsea_df <- rbind(gsea_df, as.data.frame(fgseaRes))

}

gsea_df



# select the pathways:
cur_pathways <- subset(fgseaRes, pval < 0.05)
dim(cur_pathways)

df <- gseaCurve(ranks, tf_targets$primary) #, fgseaRes)

p <- ggplot2::ggplot() + 
  geom_gsea(df) 

pdf(paste0(fig_dir, 'test_fgsea_stff2.pdf'), width=16, height=10)
p
dev.off()


#---------------------------------------------------------------------------#
# GSEA (old, just as a test)
#---------------------------------------------------------------------------#

library(fgsea)
library(gggsea)

nr4a2_targets <- subset(network_df, source == 'Nr4a2' & target_type == 'primary' & cell_group == 'MHb-Neuron' & group == 'Nurr2c')

subset(regulons, tf == 'Nr4a2') %>% arrange(-Gain) %>% dim

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes_full <- hg38_mm10_genes
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))
hg38_mm10_genes <- subset(hg38_mm10_genes, mm10_name != '' & hg38_name != '')

# need to make sure that there's only one entry for each gene in hg38_mm10_genes
mm10_genes <- unique(hg38_mm10_genes$mm10_name)
hg38_genes <- unique(hg38_mm10_genes$hg38_name)
hg38_mm10_genes <- hg38_mm10_genes[match(mm10_genes, hg38_mm10_genes$mm10_name),]

ix <- match(nr4a2_targets$target, hg38_mm10_genes$mm10_name)
nr4a2_targets$human_gene <- hg38_mm10_genes[ix,'hg38_name']
nr4a2_targets$human_gene <- ifelse(is.na(nr4a2_targets$human_gene), toupper(nr4a2_targets$target), nr4a2_targets$human_gene)

# weight 
nr4a2_targets$weight <- nr4a2_targets$Gain 
ranks <- nr4a2_targets$weight; names(ranks) <- nr4a2_targets$human_gene
ranks <- ranks[order(ranks)]


# load the GO Biological Pathways file (donwloaded from EnrichR website)
pathways <- gmtPathways('/dfs7/swaruplab/smorabit/resources/Enrichr/GO_Biological_Process_2021.txt')

# run fgsea to compute enrichments
fgseaRes <- fgsea(
  pathways = pathways, 
  stats    = ranks,
  minSize  = 15,
  maxSize  = 500
)

```






















Not used below::








Try hdWGCNA with the pseudobulk 

Not using

```{r eval=FALSE}

seurat_obj <- readRDS(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/revision/data/harmony_annotated_integration.rds')
seurat_obj <- subset(seurat_obj, cell_type %in% c('MHb-Neuron') & Group %in% c('Nurr2c', 'GFP'))


seurat_obj <- SetupForWGCNA(
  seurat_obj,
  gene_select = "fraction",
  fraction = 0.05,
  #group.by = 'annotation',
  wgcna_name = "MHb_pseudobulk"
)
length(GetWGCNAGenes(seurat_obj))


# compute pseudo-bulk replicates 
datExpr <- hdWGCNA:::to_pseudobulk(
    seurat_obj,
    replicate_col = 'Sample',
    cell_type_col = 'annotation',
    label_col = 'Group'
)
datExpr <- Reduce(cbind, lapply(names(datExpr), function(x){
    cur <- datExpr[[x]]
    colnames(cur) <- paste0(x, '_', colnames(cur))
    cur
}))
datExpr <- t(datExpr[GetWGCNAGenes(seurat_obj),])

# compute CPM:
cpm <- t(apply(datExpr, 1, function(x){
    y <- (x) / sum(x) * 1000000
    log2(y + 1)
}))
dim(cpm)

wgcna_name <- 'MHb_pseudobulk'
cur_group <- 'MHb'
cur_cpm <- cpm[grepl(cur_group, rownames(cpm)),]
seurat_obj@misc[[wgcna_name]]$datExpr <- cur_cpm



# test the soft power threshold
seurat_obj <- TestSoftPowers(seurat_obj)

plot_list <- PlotSoftPowers(seurat_obj)

#assemble with patchwork
pdf(paste0(fig_dir, 'softpower_pseudobulk2.pdf'), width=12, height=8)
wrap_plots(plot_list, ncol=2)
dev.off()


seurat_obj <- ConstructNetwork(
    seurat_obj, 
    tom_name='MHb3', 
    overwrite_tom=TRUE,
    mergeCutHeight=0.1,
)
length(unique(GetModules(seurat_obj)$module))


# plot the dendrogram
pdf(paste0(fig_dir, "dendro_pseudobulk3.pdf"),height=3, width=9)
PlotDendrogram(seurat_obj, main='MHb hdWGCNA Dendrogram')
dev.off()


seurat_obj <- ScaleData(seurat_obj, features=rownames(seurat_obj)[1:100])
seurat_obj$Assignment <- droplevels(seurat_obj$Assignment)
seurat_obj <- ModuleEigengenes(
    seurat_obj,
    group.by.vars="Assignment" # snRNAseq batch
)

# compute module connectivity:
seurat_obj <- ModuleConnectivity(
  seurat_obj,
  group.by = 'cell_type', group_name = 'MHb-Neuron'
)


modules <- GetModules(seurat_obj)
subset(modules, gene_name == 'Nr4a2')
subset(modules, gene_name == 'Gabra2')
subset(modules, gene_name == 'Sox5')
subset(modules, gene_name == 'Grin3a')
table(modules$module)



modules <- GetModules(seurat_obj)
subset(modules, gene_name == 'Nr4a2')
subset(modules, gene_name == 'Gabra2')
subset(modules, gene_name == 'Sox5')
table(modules$module)


plot_list <- ModuleFeaturePlot(
  seurat_obj, 
  order=TRUE, 
  raster=TRUE, 
  restrict_range=FALSE, 
  raster_dpi=200, 
  raster_scale=0.5, 
  point_size=1,
  reduction = 'neuron_umap'
  )
pdf(paste0(fig_dir, "MHb_featureplot_MEs.pdf"),height=8, width=16)
wrap_plots(plot_list, ncol=8)
dev.off()







########################################
# Dot Plot
########################################


MEs <- GetMEs(seurat_obj)
modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mods <- mods[mods!='grey']

seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)


# make dotplot
p <- DotPlot(
  seurat_obj,
  group.by='annotation',
  features = rev(mods)
#  scale=FALSE, col.max=5
) + coord_flip() + RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue') + xlab('') + ylab('') +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=1)
  )

pdf(paste0(fig_dir, 'MHb_hMEs_dotplot.pdf'), width=9, height=9)
p
dev.off()

# remove from seurat obj
seurat_obj@meta.data <- seurat_obj@meta.data[,!(colnames(seurat_obj@meta.data) %in% colnames(MEs))]




###################################################################################
# Run enrichr with these modules
##################################################################################


library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021','WikiPathway_2021_Human', 'KEGG_2021_Human')

# compute GO terms:
seurat_obj <- RunEnrichr(seurat_obj, dbs=dbs, max_genes=Inf)

# inspect the enrichr table:
enrichr_df <- GetEnrichrTable(seurat_obj) %>% subset(P.value < 0.05)
enrichr_df$ngenes <- unlist(lapply(strsplit(enrichr_df$Genes, ';'), function(x){length(x)}))
enrichr_df <- subset(enrichr_df, ngenes >= 3)

write.table(enrichr_df, quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'MHb_hdWGCNA_enrichr_pseudobulk.tsv'))

subset(enrichr_df, P.value < 0.05) %>% 
write.table(quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'MHb_hdWGCNA_enrichr_pseudobulk_signif.tsv'))



################################################################################
# Hubgene circle plots:
################################################################################

library(igraph)

# individual module networks
ModuleNetworkPlot(
  seurat_obj,
  mods = "all",
  #label_center=TRUE,
  outdir = paste0(fig_dir, 'MHb_hubNetworks/')
)


################################################################################
# UMAP:
################################################################################


seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 6,
  n_neighbors=10,
  min_dist=0.4,
  spread=3,
  supervised=TRUE,
  target_weight=0.3
)


# get the hub gene UMAP table from the seurat object
umap_df <- GetModuleUMAP(seurat_obj)

# plot with ggplot
p <- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(
   color=umap_df$color,
   size=umap_df$kME*2
  ) +
  umap_theme()

pdf(paste0(fig_dir, 'MHb_hubgene_umap_ggplot.pdf'), width=5, height=5)
p
dev.off()









```














Compute the module overlap with the old modules 

```{r eval=FALSE}

seurat_old <- readRDS('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/harmony_annotated_scWGCNA.rds')

################################################################################
# Module overlap analysis between Chen et al modules and modules from this study
################################################################################

library(GeneOverlap)

group1 <- 'old'
group2 <- 'new'

m1 <- GetModules(seurat_old, wgcna_name='MHb')
m2 <- GetModules(seurat_obj)

mods1 <- levels(m1$module); mods1 <- mods1[mods1 != 'grey']
mods2 <- levels(m2$module); mods2 <- mods2[mods2 != 'grey']

genome.size <- nrow(seurat_obj)


# run overlaps between module gene lists and DEG lists:
overlap_df <- do.call(rbind, lapply(mods1, function(cur_mod1){
  cur_m1_genes <- m1 %>% subset(module == cur_mod1) %>% .$gene_name
  cur_overlap_df <- do.call(rbind, lapply(mods2, function(cur_mod2){
    print(paste0(cur_mod1, ' ', cur_mod2))
    cur_m2_genes <- m2 %>% subset(module == cur_mod2) %>% .$gene_name
    cur_overlap <- testGeneOverlap(newGeneOverlap(
        cur_m1_genes,
        cur_m2_genes,
        genome.size=genome.size
    ))
    c(cur_overlap@odds.ratio, cur_overlap@pval, cur_overlap@Jaccard, length(cur_overlap@intersection))
  })) %>% as.data.frame
  colnames(cur_overlap_df) <- c('odds_ratio', 'pval', 'Jaccard', 'size_intersection')
  cur_overlap_df$m1 <- cur_mod1
  cur_overlap_df$m2 <- mods2

  # module color:
  #cur_overlap_df$color <- modules %>% subset(module == cur_mod) %>% .$color %>% unique
  cur_overlap_df
}))

overlap_df$m1 <- factor(as.character(overlap_df$m1), levels=mods1)
overlap_df$m2 <- factor(as.character(overlap_df$m2), levels=mods2)

# re-order
tmp <- unlist(lapply(mods2, function(cur_mod2){
  cur <- subset(overlap_df, m2==cur_mod2)
  which(cur$odds_ratio == max(cur$odds_ratio))
}))

overlap_df$m2 <- factor(as.character(overlap_df$m2), levels=rev(mods2[order(tmp)]))

# adjust for multiple comparisons:
overlap_df$fdr <- p.adjust(overlap_df$pval, method='fdr')

# significance level:
overlap_df$Significance <- gtools::stars.pval(overlap_df$fdr)
overlap_df$Significance <- ifelse(
  overlap_df$Significance == '.', '',
  overlap_df$Significance
)

# plot the results as a heatmap:
maxval <- 50
plot_df <- overlap_df
plot_df$odds_ratio <- ifelse(plot_df$odds_ratio > maxval, maxval, plot_df$odds_ratio)
plot_df$textcolor <- ifelse(plot_df$odds_ratio > 0.7*maxval, 'white', 'black')


p <- plot_df %>%
  ggplot(aes(x=m1, y=m2, fill=odds_ratio)) +
  geom_tile() +
  geom_text(label=plot_df$Significance, vjust = 0.72, color=plot_df$textcolor) +
  scale_fill_gradient(low='white', high='blue') +
  RotatedAxis() +
  labs(fill = 'Odds ratio') +
  theme(
    panel.border = element_rect(fill=NA, color='black', size=1),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    #axis.title.x = element_blank(),
    #axis.title.y = element_blank(),
    #axis.text.x = element_blank(),
    #axis.ticks.x = element_blank(),
  #  axis.text.y = element_blank(),
    # axis.ticks.y = element_blank(),
    plot.margin=margin(0,0,0,0)
  ) 


pdf(paste0(fig_dir, 'new_old_module_overlap2.pdf'), width=6, height=5)
p
dev.off()


```


