
```{r eval=FALSE}

# conda activate cicero
library(Seurat)
library(tidyverse)
library(cowplot)
library(Matrix)
library(viridis)
library(harmony)
library(RColorBrewer)
library(patchwork)
library(ggpubr)
library(tictoc)
library(RColorBrewer)
library(Hmisc)
library(corrplot)
library(enrichR)
library(GeneOverlap)
library(WGCNA)
library(extrafont)
library(hdWGCNA)


enableWGCNAThreads(nThreads = 8)

set.seed(2021)
colfunc <- colorRampPalette(rev(brewer.pal(11, 'Spectral' )))
theme_set(theme_cowplot())

# scp R/* hpc3:/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/
setwd("/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/scWGCNA")

# directories
data_dir <- "data/"
fig_dir <- 'figures/'


# re-load scWGCNA seurat obj
seurat_obj <- readRDS('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/harmony_annotated_scWGCNA.rds')


# re-load color scheme:
load('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/color_scheme.rda')

setwd("/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/scWGCNA_analysis")


```


Step 1: Set up data for scWGCNA, construct metacells

```{r eval=FALSE}

seurat_obj <- SetupForWGCNA(
  seurat_obj,
  gene_select = "fraction",
  fraction = 0.05,
  wgcna_name = "MHb"
)
length(GetWGCNAGenes(seurat_obj))

# construct metacells:
seurat_obj <- MetacellsByGroups(
  seurat_obj = seurat_obj,
  group.by = c("cell_type", "Sample", "Group"),
  k = 50,
  ident.group = 'cell_type'
) %>% NormalizeMetacells()


seurat_obj <- SetDatExpr(
  seurat_obj,
  group_name ="MHb-Neuron",
  group.by = "cell_type"
)

# test the soft power threshold
seurat_obj <- TestSoftPowers(
  seurat_obj,
  group.by='cell_type',
  group_name="MHb-Neuron",
  setDatExpr=TRUE
)

# construct the network
seurat_obj <- ConstructNetwork(seurat_obj,soft_power = 6,setDatExpr = FALSE)

# plot the dendrogram
pdf(paste0(fig_dir, "MHb_dendro.pdf"),height=3, width=6)
PlotDendrogram(seurat_obj, main='MHb scWGCNA Dendrogram')
dev.off()

#################################################################################
#Compute Module eigengenes, connectivity
#################################################################################

# scale the expression matrix for all the WGCNA genes:
seurat_obj <- Seurat::ScaleData(seurat_obj,features = GetWGCNAGenes(seurat_obj))

# compute all MEs in the full single-cell dataset
seurat_obj <- ModuleEigengenes(
  seurat_obj,
  group.by.vars="Assignment" # snRNAseq batch
)

# compute module connectivity:
seurat_obj <- ModuleConnectivity(
  seurat_obj,
  group.by = 'cell_type', group_name = 'MHb-Neuron'
)


################################################################################
# Change module colors
################################################################################


library(MetBrewer)

modules <- GetModules(seurat_obj)
mods <- levels(modules$module)
mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct %>% arrange(module) %>% .$color
n_colors <- length(mod_colors) -1

new_colors <- paste0(met.brewer("Cross", n=n_colors, type='discrete'))
seurat_obj <- ResetModuleColors(seurat_obj, new_colors)


# save data:
saveRDS(seurat_obj, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/harmony_annotated_scWGCNA.rds')


########################################
# Feature Plots
########################################

plot_list <- ModuleFeaturePlot(seurat_obj, order=TRUE, raster=TRUE, restrict_range=FALSE, raster_dpi=200, raster_scale=0.5, point_size=1)
pdf("figures/MHb_featureplot_hMEs.pdf",height=12, width=6)
wrap_plots(plot_list, ncol=2)
dev.off()


################################################################################
# Hubgene circle plots:
################################################################################

library(igraph)

# individual module networks
ModuleNetworkPlot(
  seurat_obj,
  mods = "all",
  #label_center=TRUE,
  outdir = paste0(fig_dir, 'MHb_hubNetworks/')
)


################################################################################
# UMAP:
################################################################################


seurat_obj <- RunModuleUMAP(
  seurat_obj,
  n_hubs = 10,
  n_neighbors=15,
  min_dist=0.5,
  spread=3,
  supervised=TRUE,
  target_weight=0.2
)


label_genes <- c(
  'Cacna1g',
  'Csmd3',
  'Kirrel3',
  'Rapgef4',
  'Ywhae',
  'Homer2',
  'Runx1t1',
  'Cntn4',
  'Grin3a',
  'H3f3b',
  'Gabra1',
  'Gabra2',
  'Kctd8',
  'Nr4a2',
  'Pbx3',
  'Klf9',
  'Nfib',
  'Scrt1',
  'Sox10',
  'Tcf4',
  'Zbtb7a'
)


library(reshape2)
library(igraph)
pdf(paste0(fig_dir, 'MHb_hubgene_umap_igraph.pdf'), width=7, height=7)
ModuleUMAPPlot(
  seurat_obj,
  edge.alpha=0.25,
  sample_edges=TRUE,
  edge_prop=0.15,
  label_genes = label_genes,
  keep_grey_edges=FALSE,
  label_hubs=5
)
dev.off()

################################################################################
# Pathway enrichment
################################################################################

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021', 'WikiPathways_2019_Mouse', 'KEGG_2019_Mouse')

# compute GO terms:
seurat_obj <- RunEnrichr(seurat_obj, dbs=dbs, max_genes=300)
enrichr_df <- subset(enrichr_df, P.value < 0.05)
write.table(enrichr_df, quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'SupplementaryTable_MHb_GO_terms.tsv'))


modules <- GetModules(seurat_obj)
color_df <- modules %>% subset(module!='grey') %>%
  select(c(module, color)) %>% distinct %>%
  rename(c(group=module, colour=color))


# helper function to wrap text
wrapText <- function(x, len) {
    sapply(x, function(y) paste(strwrap(y, len), collapse = "\n"), USE.NAMES = FALSE)
}

combined_output <- GetEnrichrTable(seurat_obj)
selected_terms <- read.table('data/MHb_selected_GO_terms.txt', sep='\t', header=1)

# subset selected terms
selected_terms <- subset(combined_output, Term %in% selected_terms$Term & P.value < 0.05)

selected_terms$group <- factor(
  as.character(selected_terms$module),
  levels = mods
)


max_p <- 15

selected_terms$logp <- -log(selected_terms$P.value)
selected_terms$logp <- ifelse(selected_terms$logp > max_p, max_p, selected_terms$logp)
# remove GO Term ID
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")

selected_terms <- selected_terms %>%
  arrange(group)


selected_terms$wrap <- wrapText(selected_terms$Term, 35)

selected_terms$Term <- factor(
  as.character(selected_terms$Term),
  levels = rev(unique(as.character(selected_terms$Term)))
)

# GO Term dot plot
p <- selected_terms %>%
  ggplot(aes(x = group, y = Term, color =logp, size=log(Combined.Score))) +
  geom_point() +
  scale_color_stepsn(colors=rev(magma(256))) +
  RotatedAxis() + xlab('') + ylab('') +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.margin = margin(0,0,0,0),
    panel.grid = element_line(size=0.25, color='lightgrey')
  )


  # make the colorbar as its own heatmap
  color_df$var <- 1
  colorbar <- color_df %>%
    ggplot(aes(x=group, y=var, fill=group)) +
    geom_tile() +
    scale_fill_manual(values=color_df$colour) +
    coord_equal() +
    NoLegend() + RotatedAxis() +
    theme(
      plot.title=element_blank(),
      axis.line=element_blank(),
      axis.ticks.y =element_blank(),
      axis.text.y = element_blank(),
      axis.title = element_blank(),
      plot.margin=margin(0,0,0,0),
    )




pdf(paste0(fig_dir, 'selected_GO_terms.pdf'), width=8, height=8)
p  / colorbar
dev.off()

################################################################################
# Motif analysis
################################################################################

library(JASPAR2020)
library(motifmatchr)
library(TFBSTools)
library(EnsDb.Mmusculus.v79)
library(GenomicRanges)

# get the pfm from JASPAR2020 using TFBSTools
pfm_core <- TFBSTools::getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# run the motif scan with these settings for the mouse dataset
seurat_obj <- MotifScan(
  seurat_obj,
  species_genome = 'mm10',
  pfm = pfm_core,
  EnsDb = EnsDb.Mmusculus.v79
)

target_genes <- GetMotifTargets(seurat_obj)
target_genes$NR4A2

# overlap between moduls & TF target genes:
seurat_obj<- OverlapModulesMotifs(seurat_obj)

pdf(paste0(fig_dir, 'nr4a2_overlap_net.pdf'), width=6, height=6)
ModuleTFNetwork(
  seurat_obj,
  edge.alpha=0.25,
  tf_name = "NR4A2",
  tf_gene_name = "Nr4a2"
)
dev.off()


saveRDS(seurat_obj, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/harmony_annotated_scWGCNA.rds')

################################################################################
# Nr4a2 TF circuit
################################################################################

# Get info from hdWGCNA object
modules <- GetModules(seurat_obj)
pfm <- GetPFMList(seurat_obj)
motif_df <- GetMotifs(seurat_obj)
target_genes <- GetMotifTargets(seurat_obj)
umap_df <- GetModuleUMAP(seurat_obj)

# format motif dataframe so it has mouse gene names
motif_df$motif_name_orig <- motif_df$motif_name
tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
tmp_ids <- motif_df$motif_ID[motif_df$motif_name_orig %in% tmp]

motif_df <- motif_df %>%  dplyr::select(c(motif_ID, motif_name_orig, motif_name))

motif_df$motif_name <- gsub("\\s*\\([^\\)]+\\)","", motif_df$motif_name_orig)
tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
motif_df <- subset(motif_df, !(motif_name %in% tmp))

tmp2 <- strsplit(tmp, '::')
names(tmp2) <- tmp
tmp <- do.call(rbind, lapply(1:length(tmp2), function(i){
  data.frame(motif_ID = tmp_ids[i], motif_name_orig = names(tmp2)[i], motif_name = unlist(as.character(tmp2[[i]])))
}))
motif_df <- rbind(motif_df, tmp)

# convert gene names to mouse style
motif_df$motif_name <- str_to_title(motif_df$motif_name)

# Which TFs regulate Nr4a2?
cur_tfs <- unlist(lapply(names(target_genes), function(x){"Nr4a2" %in% target_genes[[x]]}))
cur_tfs <- names(target_genes)[cur_tfs]
cur_tfs <- motif_df %>% subset(motif_name_orig %in% cur_tfs) %>% .$motif_name
cur_tfs <- subset(umap_df, gene %in% c(cur_tfs, 'Nr4a2'))
cur_tfs


# Loop through each TF so we can make the V & E dataframes for an igraph object
tf_list <- unique(cur_tfs$gene)
cur_motif_df <- subset(motif_df, motif_name %in% tf_list)

link_df <- data.frame()
for(cur_tf in tf_list){
  print(cur_tf)
  cur_tf_id <- subset(motif_df, motif_name == cur_tf) %>% .$motif_name_orig %>% head(1)
  cur_targets <- target_genes[[cur_tf_id]]
  cur_targets <- cur_targets[cur_targets %in% tf_list]

  cur_links <- data.frame(
    source = cur_tf,
    target = cur_targets,
    value =  as.numeric(TOM[cur_tf, cur_targets])
  )
  link_df <- rbind(cur_links, link_df)

}

# replace 0 value with max (self loops)
link_df$value <- ifelse(link_df$value == 0, max(link_df$value), link_df$value)
plot_lim <- 0.2
link_df$value <- ifelse(link_df$value > plot_lim, plot_lim, link_df$value)

high_color = celltype_colors['MHb-Neuron']
low_color = 'grey90'
# make a dummy ggplot so I can get the colors:
plot_df <- link_df
plot_df$source <- factor(as.character(plot_df$source), levels=rev(c('Nr4a2', 'Klf9', 'Nfib', 'Scrt1', 'Sox10', 'Tcf4', 'Zbtb7a')))

plot_df$target <- factor(as.character(plot_df$target), levels=rev(c('Nr4a2', 'Klf9', 'Nfib', 'Scrt1', 'Sox10', 'Tcf4', 'Zbtb7a')))


p <- ggplot(plot_df, aes(x=forcats::fct_rev(source), y=target, color=value, size=value)) +
  geom_point() +
  scale_color_gradient(high=high_color,  low=low_color,limits=c(0, plot_lim)) +
  RotatedAxis() + theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(fill=NA, color='black', size=1),
    axis.text = element_text(face='italic')
  ) + coord_equal() + xlab('Source TF') + ylab("Target TF")


ggp <- ggplot_build(p)
link_df$color <- ggp$data[[1]]$colour

pdf(paste0(fig_dir, 'nr4a2_tf_circuit_graph_legend.pdf'), width=4, height=4, useDingbats=FALSE)
p
dev.off()



v_df <- cur_tfs %>% dplyr::select(c(gene, module, color, UMAP1, UMAP2, kME))

# make graph:
g <- igraph::graph_from_data_frame(
  link_df,
  directed=TRUE,
  vertices=v_df
)

lay <- layout_with_graphopt(
  g,
  start = as.matrix(v_df[,c('UMAP1', 'UMAP2')]),
  niter=3
)


pdf(paste0(fig_dir, 'nr4a2_tf_circuit_graph.pdf'), width=5, height=5, useDingbats=FALSE)

plot(
  g,
  layout=lay,
  edge.color=adjustcolor(E(g)$color),
  vertex.size=15,
  edge.curved=0.2,
  edge.width=E(g)$value*10,
  edge.arrow.size=0.5,
  vertex.color=V(g)$color,
  vertex.label=V(g)$gene,
  vertex.label.dist=1.1,
  vertex.label.degree=-pi/4,
  vertex.label.family='Helvetica',
  vertex.label.font = 3,
  vertex.label.color = 'black',
  vertex.label.cex=0,
  vertex.frame.color='black',
  margin=0
)

dev.off()

################################################################################
# Plot the overlap between NR4A2 target genes & as baplot
################################################################################

df <- GetMotifOverlap(seurat_obj)

cur_df <- df %>% subset(tf == 'NR4A2')

plot_var <- 'odds_ratio'
p <- cur_df %>%
  ggplot(aes(y=reorder(module, odds_ratio), x=odds_ratio)) +
  geom_bar(stat='identity', fill=cur_df$color) +
  geom_vline(xintercept = 1, linetype='dashed', color='gray') +
  geom_text(aes(label=Significance), color='black', size=3.5, hjust='center') +
  ylab('') +
  xlab("Odds Ratio") +
  theme(
    plot.title = element_text(hjust = 0.5)
  )


pdf(paste0(fig_dir, 'Nr4a2_motif_overlap_or2.pdf'), width=3, height=3, useDingbats=FALSE)
p
dev.off()


modules <- GetModules(seurat_obj)
module_colors <- modules %>% subset(module != 'grey') %>% select(c(module, color)) %>% distinct
cp <- module_colors$color; names(cp) <- as.character(module_colors$module)

cur_df <- arrange(cur_df, odds_ratio)
cur_df$module <- factor(as.character(cur_df$module), levels=as.character(cur_df$module))

cur_df$shape <- ifelse(cur_df$fdr < 0.05, 21, 4)

p <- cur_df %>%
  ggplot(aes(y=reorder(module, odds_ratio), x=odds_ratio, size=-log(fdr), fill=module, color=module)) +
  geom_vline(xintercept = 1, linetype='dashed', color='lightgrey') +
  geom_segment(aes(y=module, yend=module, x=0, xend=odds_ratio), size=0.5, alpha=0.5) +
  geom_point(shape=cur_df$shape, color='black') +
  scale_color_manual(values=cp) +
  scale_fill_manual(values=cp) +
  ylab('') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 2.5), breaks=c(0,1,2)) +
  xlab("Odds Ratio")
#  scale_fill_gradient(high = 'red', low='white')

pdf(paste0(fig_dir, 'nr4a2_deg_overlap_lollipop.pdf'), width=4, height=3)
p
dev.off()

```

Plot Nr4a2 motif logo

```{r eval=FALSE}


  # get pfm list from Seurat obj
  pfm <- GetPFMList(seurat_obj)
  motif_df <- GetMotifs(seurat_obj)

  cur_id <- subset(motif_df, motif_name == 'NR4A2') %>% .$motif_ID

  p <- ggplot() +
   ggseqlogo::geom_logo( as.matrix(pfm[[cur_id]])) +
   ggseqlogo::theme_logo() +
   xlab('') + ylab('') + theme(
     axis.text.x=element_blank(),
     axis.text.y=element_blank(),
     axis.title.y = element_text(angle=0),
     plot.margin = margin(t = 0,  # Top margin
                            r = 0,  # Right margin
                            b = 0,  # Bottom margin
                            l = 0) # Left margin
   )


pdf(paste0(fig_dir, 'nr4a2_motif_logo.pdf'), width=3, height=3)
p
dev.off()

```

Compare the eigengenes between Nurr2c & GFP:

```{r eval=FALSE}

# seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)
MEs <- GetMEs(seurat_obj)
mods <- colnames(MEs); mods <- mods[mods != 'grey']
seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)

p <- custom_vln(
    seurat_obj,
    features = mods,
    group.by = 'annotation',
    split.by = 'Group',
    groups = c('MHb-1', 'MHb-2', 'MHb-3', 'MHb-4', 'MHb-5'),
    add_boxplot=FALSE,
    split_colors=c('darkgoldenrod3', 'hotpink3'),
    add_colorbar=FALSE,
    plot_ymin = NA
  )

pdf(paste0(fig_dir, 'MHb_hME_vln_stack.pdf'), width=6, height=8)
p
dev.off()


```

Supp tables:

```{r eval=FALSE}

# Nr4a2 targets:
target_genes <- GetMotifTargets(seurat_obj)
nr4a2_targets <- target_genes$NR4A2

nr4a2_df <- data.frame(
  target = as.character(nr4a2_targets)
)

write.csv(nr4a2_df, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/SupplementaryTable_nr4a2_targets.csv', quote=FALSE, row.names=FALSE)

degs <- read.csv(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/cluster_marker_DEGs.csv')

degs$Nr4a2 <- degs$gene %in% nr4a2_targets

write.csv(degs, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/SupplementaryTable_cluster_marker_DEGs.csv', quote=FALSE, row.names=FALSE)

degs <- read.csv(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/cluster_Nurr2c_vs_GFP.csv')
degs$Nr4a2 <- degs$gene %in% nr4a2_targets

degs_ct <- read.csv(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/celltype_Nurr2c_vs_GFP.csv')
degs_ct$Nr4a2 <- degs_ct$gene %in% nr4a2_targets

degs_ct <- subset(degs_ct, group %in% c("MHb-Neuron", "LHb-Neuron", "PHb-Neuron"))

degs <- rbind(degs_ct, degs)

write.csv(degs, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/SupplementaryTable_Nurr2c_vs_GFP.csv', quote=FALSE, row.names=FALSE)

markers1 <- read.csv(file="/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/mhb_markers.csv")
markers2 <- read.csv(file="/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/lhb_markers.csv")
markers3 <- read.csv(file="/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/data/phb_markers.csv")
markers3 <- markers3 %>% dplyr::select(-X.1)

markers <- rbind(markers1, markers2, markers3) %>% dplyr::select(-X)

markers$Nr4a2 <- markers$gene %in% nr4a2_targets

write.csv(markers, '/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/SupplementaryTable_neuron_markers.csv', quote=FALSE, row.names=FALSE)


```



Compute the overlap between Nurr2c vs GFP DEGs and the Nr4a2 target genes:

```{r eval=FALSE}

library(GeneOverlap)

# load degs:
degs <- read.csv(file='/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/Nurr2c_vs_GFP/DEGs/data/celltype_Nurr2c_vs_GFP.csv')
degs$group <- factor(
  degs$group,
  levels = levels(seurat_obj$cell_type)
)
ident1 <- unique(degs$ident1)
ident2 <- unique(degs$ident2)

# Nr4a2 targets:
target_genes <- GetMotifTargets(seurat_obj)
cur_targets <- target_genes$NR4A2

# settings
genome.size <- nrow(seurat_obj)
cell_groups <- levels(degs$group)
fc_cutoff <- 0.25

overlap_df <- do.call(rbind, lapply(cell_groups, function(cur_group){
  #print(cur_group)
  cur_DEGs_up <- degs %>% subset(group == cur_group & p_val_adj <= 0.05 & avg_log2FC > fc_cutoff) %>% .$gene
  cur_DEGs_down <- degs %>% subset(group == cur_group & p_val_adj <= 0.05 & avg_log2FC < -1*fc_cutoff) %>% .$gene

  cur_overlap_up <- testGeneOverlap(newGeneOverlap(
      cur_targets,
      cur_DEGs_up,
      genome.size=genome.size
  ))
  cur_overlap_down <- testGeneOverlap(newGeneOverlap(
      cur_targets,
      cur_DEGs_down,
      genome.size=genome.size
  ))

  cur_overlap <- data.frame(
    'odds.ratio' = c(cur_overlap_up@odds.ratio, cur_overlap_down@odds.ratio),
    'pval' = c(cur_overlap_up@pval, cur_overlap_down@pval),
    'Jaccard' = c(cur_overlap_up@Jaccard, cur_overlap_down@Jaccard),
    'size_intersection' = c(length(cur_overlap_up@intersection), length(cur_overlap_down@intersection)),
    'group' = c(cur_group, cur_group),
    'condition' = c(ident1, ident2)
  )

  cur_overlap
})) %>% as.data.frame


################################################################################
# Plot as a lollipop
################################################################################

overlap_df$group_condition <- paste0(overlap_df$group, '_', overlap_df$condition)
overlap_df$condition <- factor(as.character(overlap_df$condition, levels=c(ident1, ident2)))

# remove unwanted groups
exclude_groups <- c('EPD', 'END', 'PER', 'OPC')
overlap_df <- overlap_df %>% subset(!(group %in% exclude_groups))

overlap_df$group_condition <- factor(
  as.character(overlap_df$group_condition),
  levels = as.character(overlap_df$group_condition)
)

overlap_df$group_condition <- fct_rev(overlap_df$group_condition)

overlap_df$shape <- ifelse(overlap_df$pval < 0.05, 21, 4)
geom_point() +


p <- overlap_df %>%
  ggplot(aes(y=group_condition, x=odds.ratio, size= size_intersection, color=condition)) +
#  geom_vline(xintercept = 1, linetype='dashed', color='gray') +
  geom_segment(aes(y=group_condition, yend=group_condition, x=0, xend=odds.ratio), size=0.5, color='grey') +
  geom_point() +
  scale_size(range=c(2,7)) +
  scale_color_manual(values=c('hotpink3', 'darkgoldenrod3')) +
  ylab('') + xlab("Odds ratio") +
  scale_x_continuous(breaks = c(0, 1, 2))
#  scale_fill_gradient(high = 'red', low='white')

pdf(paste0(fig_dir, 'test_nr4a2_deg_overlap.pdf'), width=5, height=2.5)
p
dev.off()

p <- overlap_df %>%
  ggplot(aes(y=group_condition, x=odds.ratio, size= size_intersection, color=condition, fill=condition)) +
#  geom_vline(xintercept = 1, linetype='dashed', color='gray') +
  geom_segment(aes(y=group_condition, yend=group_condition, x=0, xend=odds.ratio), size=0.5, color='grey') +
  geom_point(shape=overlap_df$shape, color='black') +
  scale_size(range=c(2,7)) +
  scale_color_manual(values=c('hotpink3', 'darkgoldenrod3')) +
  ylab('') + xlab("Odds ratio") +
  scale_x_continuous(breaks = c(0, 1, 2))
#  scale_fill_gradient(high = 'red', low='white')

pdf(paste0(fig_dir, 'test_nr4a2_deg_overlap_sign.pdf'), width=5, height=3)
p
dev.off()



```


Project Modules to Hashikawa dataset:

```{r eval=FALSE}

# re-load processed hashikawa dataset:
seurat_hashikawa <- readRDS(file='data/hashikawa_scWGCNA.rds')


seurat_hashikawa <- readRDS('/dfs7/swaruplab/smorabit/collab/woodlab/cocaine_mouse_2021/analysis/test_harmony/data/hashikawa_seurat.rds')

seurat_hashikawa$cell_type <- ifelse(grepl("MHb", seurat_hashikawa$celltype_neuron), 'MHb', seurat_hashikawa$celltype_neuron)

seurat_hashikawa <- ProjectModules(
  seurat_obj = seurat_hashikawa,
  seurat_ref = seurat_obj,
  scale_genes = TRUE,
  wgcna_name_proj="MHb_projected",
  wgcna_name = "MHb"
)

# plot modules in hashikawa data:
plot_list <- ModuleFeaturePlot(
  seurat_hashikawa, order="shuffle", raster=TRUE,
  restrict_range=FALSE, point_size=0.5, raster_scale=0.5
)
pdf("figures/MHb_featureplot_hMEs_hashikawa.pdf",height=6, width=12)
wrap_plots(plot_list, ncol=4)
dev.off()


# remove legend, title, then plot
for(i in 1:length(plot_list)){
  plot_list[[i]] <- plot_list[[i]] + NoLegend() + ggtitle('') + theme(plot.margin = margin(0,0,0,0))
}

png("figures/MHb_featureplot_hMEs_hashikawa.png",height=6, width=12, units='in', res=1000)
wrap_plots(plot_list, ncol=4)
dev.off()

p2 <- DimPlot(
  seurat_hashikawa, group.by='celltype_neuron',
  label = TRUE) +
  #scale_color_manual(values=cluster_colors) +
  ggtitle('') +
  umap_theme() + NoLegend()



pdf("figures/hashikawa_umap.pdf",height=5, width=5)
p2
dev.off()


################################################################################
# Module preservation
################################################################################


seurat_hashikawa <- SetDatExpr(
  seurat_hashikawa,
  group_name = "MHb",
  group.by = "cell_type"
)

# run module preservation function
seurat_hashikawa <- ModulePreservation(
  seurat_hashikawa,
  seurat_ref = seurat_obj,
  name="MHb_projected",
  verbose=3
)

plot_list <- PlotModulePreservation(
  seurat_hashikawa,
  name="MHb_projected",
  statistics = "summary"
)

pdf(paste0(fig_dir, 'hashikawa_module_preservation_summary.pdf'), width=10, height=5)
wrap_plots(plot_list, ncol=2)
dev.off()


saveRDS(seurat_hashikawa, file='data/hashikawa_scWGCNA.rds')
seurat_hashikawa <- readRDS(file='data/hashikawa_scWGCNA.rds')

```
